torch/distributed/optim/functional_adam.py

```diff
@@ -45,13 +46,62 @@ def __init__(
         self.amsgrad = amsgrad
         self.state = torch.jit.annotate(Dict[torch.Tensor, Dict[str, torch.Tensor]], {})
 
-        if len(params) == 0:
+        if len(params) == 0 and not allow_empty_param_list:
             raise ValueError("optimizer got an empty parameter list")
 
         # NOTE: we only have one param_group and don't allow user to add additional
         # param group as it's not a common use case.
         self.param_group = {"params": params}
 
+    def step_param(self, param: Tensor, grad: Optional[Tensor]):
+        """
+        Similar to step, but operates on a single parameter and optionally a
+        gradient tensor.
+        """
+        params = [param]
+        params_with_grad = []
+        grads = []
+        exp_avgs = []
+        exp_avg_sqs = []
+        max_exp_avg_sqs = []
+        state_steps: List[int] = []
+        if grad is not None:
+            params_with_grad.append(param)
+            grads.append(grad)
+        if param not in self.state:
+            self.state[param] = {}
+            state = self.state[param]
+            state['step'] = torch.tensor(0.0)
+            state['exp_avg'] = torch.zeros_like(param, memory_format=torch.preserve_format)
+            state['exp_avg_sq'] = torch.zeros_like(param, memory_format=torch.preserve_format)
+            if self.amsgrad:
+                state['max_exp_avg_sq'] = torch.zeros_like(param, memory_format=torch.preserve_format)
```

Comment:
After this line, will `self.state[param]` be updated as well? If not, will `state` be overwritten as an empty dict in L80?

Reply:
Yeah, we can think of `self.state[param]` as a dict of dicts. For example in the following program:
```
d = {}
if 1 not in d:
    d[1] = {}
    state = d[1]
    state[1] = 1

state = d[1]
print(state)
```
gives
` >>> {1: 1}`
as expected

The same thing is also done in `step()` function in this file (we have an issue to dedupe the implementations as a bootcamp task).

Reply:
Thanks for the explanation!